### ISSUE TITLE
Rotate Recovery behavior not sufficient 

### ISSUE THREAD

**Issue by vinnnyr** [ros-agriculture/lawn_tractor#15]
The current Move Base / Move Base Flex Recovery Behavior (the default Rotate Recovery) does not work to get the lawn tractor 'unstuck' from a collision into the simulation wall. 

I have confirmed this behavior does not work as expected even before the move base flex SMACH state machine implementation (Commit 4e2f6eaae58020a64982d82cd812933c36a73499)

Any help on this issue will be much appreciated.

**Comment by Grace-Hamilton** [comment]
Would you mind if I worked on this?

**Comment by Grace-Hamilton** [comment]
Hey @vinnnyr , just wanted to clarify something regarding the `mbf_state_machine.py` file (src in lawn_tractor_navigation). With regards to the `recovery_path_goal_cb` method, there is this one block of code in lines 95-109:
```
smach.StateMachine.add(
            'RECOVERY',
             smach_ros.SimpleActionState(
                'move_base_flex/recovery',
                RecoveryAction,
                goal_cb=recovery_path_goal_cb,
                input_keys=["error", "clear_costmap_flag"],
                output_keys = ["error_status", 'clear_costmap_flag']
            ),
            transitions={
                'succeeded': 'GET_PATH',
                'aborted': 'aborted',
                'preempted': 'preempted'
            }
        )

```
What exactly is `RecoveryAction` here and what does it do/which file is it set up in? I can't seem to find anything else about it in any of the other files.

**Comment by vinnnyr** [comment]
Recovery Action is from Move Base Flex messages, found here: https://github.com/ros-agriculture/lawn_tractor/blob/5e8762c310b1a1f9d349f168ccb5a0750ab7c18d/lawn_tractor_navigation/src/mbf_state_machine.py#L13

This is the message type that is interacting with the Move Base Flex Action Server. The recovery behavior that the robot is currently using is not working 

http://docs.ros.org/kinetic/api/mbf_msgs/html/action/Recovery.html


Please let me know if this helps, but basically the task is to find a recovery behavior that works well for our application. 

By the way, I am currently working on moving from this SMACH implementation to BehaviorTree.cpp (See Issue #13 ), but if you are more comfortable with the SMACH implementation, I suggest you either fork the repo or checkout your own branch before I merge the BehaviorTree implementation into master. 

The lack of proper recovery behavior is still a hot issue that we need help on

**Comment by Grace-Hamilton** [comment]
So just to clarify, Recovery Action is a type of message? Is there code anywhere in this repository that directly dictates how the robot moves during recovery?

**Comment by vinnnyr** [comment]
Yes, I guess Recovery Action is defined is a service message _(not sure if that is precisely the correct term)_ http://wiki.ros.org/actionlib?distro=noetic

Essentially Move Base Flex utilizes the actionlib which utilizes ROS services.,

To answer your second question — The recovery behaviors are listed in lawn_tractor_navigation/config/recovery_behaviors.yaml, which is being sent to the ROS param server in the simulation launch file : https://github.com/ros-agriculture/lawn_tractor/blob/master/lawn_tractor_sim/launch/lawn_tractor_sim.launch


I hope this helps, let me know if I can further clarify anything :)

**Comment by droter** [comment]
Car-like recovery behavior:  https://github.com/gkouros/car-maneuver-recovery

Might be good to add to the sim?
